File: src/mcp_rtt_server/tools.py
"""MCP Tools for Segger RTT via J-Link."""

from __future__ import annotations

import asyncio
import json

from mcp.types import Tool, TextContent

from .jlink_rtt import get_instance


def get_tool_definitions() -> list[Tool]:
    """Get all MCP tool definitions."""
    return [
        Tool(
            name="jlink_connect",
            description="Connect to J-Link debugger and start RTT monitoring. "
                        "Must be called before reading or writing RTT data.",
            inputSchema={
                "type": "object",
                "properties": {
                    "serial": {
                        "type": "string",
                        "description": "J-Link serial number (optional). "
                                       "If not provided, connects to first available J-Link.",
                    },
                    "device": {
                        "type": "string",
                        "description": "Target device name (default: HC32L19x). "
                                       "Examples: Cortex-M0+, STM32F103, etc.",
                        "default": "HC32L19x",
                    },
                    "speed": {
                        "type": "integer",
                        "description": "JTAG/SWD speed in kHz (default: 4000).",
                        "default": 4000,
                    },
                },
            },
        ),
        Tool(
            name="jlink_disconnect",
            description="Disconnect from J-Link debugger and stop RTT monitoring.",
            inputSchema={
                "type": "object",
                "properties": {},
            },
        ),
        Tool(
            name="rtt_read",
            description="Read accumulated RTT data from the ring buffer. "
                        "Data is continuously collected by the background monitor thread. "
                        "Use this to get output from the target device.",
            inputSchema={
                "type": "object",
                "properties": {
                    "channel": {
                        "type": "integer",
                        "description": "RTT channel to read from (default: 0).",
                        "default": 0,
                    },
                    "max_bytes": {
                        "type": "integer",
                        "description": "Maximum bytes to return (default: 512). "
                                       "Note: This reads from the accumulated buffer.",
                        "default": 512,
                    },
                },
            },
        ),
        Tool(
            name="rtt_read_log",
            description="Read the tail of the complete RTT log file. This is an "
                        "independent broadcast sink (not drained by other clients), "
                        "so it returns the full RTT output even while another client "
                        "(e.g. the VSCode extension) is streaming via rtt_read. "
                        "Prefer this over rtt_read when another consumer is active.",
            inputSchema={
                "type": "object",
                "properties": {
                    "max_bytes": {
                        "type": "integer",
                        "description": "Maximum bytes to return from the end of the log (default: 8192).",
                        "default": 8192,
                    },
                },
            },
        ),
        Tool(
            name="rtt_read_raw",
            description="Read new bytes from the broadcast log starting at a byte offset. "
                        "Non-draining and multi-consumer safe: ideal for a continuous "
                        "monitor that must coexist with other readers without stealing "
                        "their data. Pass the returned next_offset as 'offset' on the "
                        "next call; if the log rotated (next_offset > file size), pass "
                        "offset=0. Returns JSON: {\"data\": \"...\", \"next_offset\": N}.",
            inputSchema={
                "type": "object",
                "properties": {
                    "offset": {
                        "type": "integer",
                        "description": "Byte offset to read from (default: 0).",
                        "default": 0,
                    },
                    "max_bytes": {
                        "type": "integer",
                        "description": "Maximum bytes to return (default: 8192).",
                        "default": 8192,
                    },
                },
            },
        ),
        Tool(
            name="rtt_write",
            description="Write data to RTT down-buffer (host -> device). "
                        "The target device must be running RTT with a down-buffer listener.",
            inputSchema={
                "type": "object",
                "properties": {
                    "channel": {
                        "type": "integer",
                        "description": "RTT channel to write to (default: 0).",
                        "default": 0,
                    },
                    "data": {
                        "type": "string",
                        "description": "String data to send to the target device.",
                    },
                },
                "required": ["data"],
            },
        ),
        Tool(
            name="rtt_list_devices",
            description="List available J-Link debuggers connected to the host.",
            inputSchema={
                "type": "object",
                "properties": {},
            },
        ),
        Tool(
            name="jlink_status",
            description="Get current J-Link connection status and RTT buffer information.",
            inputSchema={
                "type": "object",
                "properties": {},
            },
        ),
        Tool(
            name="rtt_clear",
            description="Clear the RTT ring buffer. "
                        "This does not clear the actual RTT buffers on the device.",
            inputSchema={
                "type": "object",
                "properties": {},
            },
        ),
    ]


async def call_tool(name: str, arguments: dict) -> list[TextContent]:
    """Call a tool by name with the given arguments.

    The JLinkRTT layer wraps pylink, which performs blocking USB/DLL I/O. To keep
    the SSE daemon's asyncio event loop responsive for all clients, every blocking
    pylink call is offloaded to a worker thread via ``asyncio.to_thread``. The
    stdio server (single client) is unaffected by this.

    Args:
        name: Tool name.
        arguments: Tool arguments.

    Returns:
        List of TextContent results.
    """
    rtt = get_instance()

    if name == "jlink_connect":
        serial = arguments.get("serial")
        device = arguments.get("device", "HC32L19x")
        speed = arguments.get("speed", 4000)

        # Shared daemon: if a connection already exists (another client opened it),
        # keep it. Force-reconnecting here would drop the J-Link for everyone.
        if rtt.is_connected():
            status = rtt.status()
            return [TextContent(
                type="text",
                text=f"Already connected to J-Link device '{status.device_name}' "
                     f"(serial: {status.serial}, speed: {status.speed} kHz)\n"
                     f"RTT monitoring active on channel {status.channel} (shared connection)",
            )]

        success, err_msg = await asyncio.to_thread(
            rtt.connect, serial=serial, device=device, speed=speed
        )
        if success:
            status = rtt.status()
            return [TextContent(
                type="text",
                text=f"Connected to J-Link device '{status.device_name}' "
                     f"(serial: {status.serial}, speed: {status.speed} kHz)\n"
                     f"RTT monitoring started on channel {status.channel}",
            )]
        else:
            return [TextContent(
                type="text",
                text=f"Failed to connect to J-Link.\n\nError:\n{err_msg}",
            )]

    elif name == "jlink_disconnect":
        if not rtt.is_connected():
            return [TextContent(type="text", text="J-Link is not connected.")]

        await asyncio.to_thread(rtt.disconnect)
        return [TextContent(type="text", text="J-Link disconnected successfully.")]

    elif name == "rtt_read":
        if not rtt.is_connected():
            return [TextContent(
                type="text",
                text="J-Link is not connected. Call jlink_connect first.",
            )]

        channel = arguments.get("channel", 0)
        max_bytes = arguments.get("max_bytes", 512)

        # NOTE: the monitor thread is the sole reader of the physical RTT up-buffer.
        # rtt_read returns only what it accumulated in the ring buffer, so it never
        # steals bytes from the monitor's log file / stderr / other consumers.
        ring_data = await asyncio.to_thread(rtt.read, channel=channel, max_bytes=max_bytes)
        return [TextContent(type="text", text=ring_data if ring_data else "(no RTT data)")]

    elif name == "rtt_read_log":
        max_bytes = arguments.get("max_bytes", 8192)
        data = await asyncio.to_thread(rtt.read_log_tail, max_bytes)
        return [TextContent(type="text", text=data if data else "(no RTT log yet — connect first)")]

    elif name == "rtt_read_raw":
        offset = arguments.get("offset", 0)
        max_bytes = arguments.get("max_bytes", 8192)
        data, next_offset = await asyncio.to_thread(rtt.read_log_raw, offset, max_bytes)
        payload = json.dumps({"data": data, "next_offset": next_offset})
        return [TextContent(type="text", text=payload)]

    elif name == "rtt_write":
        if not rtt.is_connected():
            return [TextContent(
                type="text",
                text="J-Link is not connected. Call jlink_connect first.",
            )]

        channel = arguments.get("channel", 0)
        data = arguments.get("data", "")

        if not data:
            return [TextContent(type="text", text="No data provided to write.")]

        written = await asyncio.to_thread(rtt.write, channel=channel, data=data)
        if written >= 0:
            return [TextContent(
                type="text",
                text=f"Wrote {written} bytes to RTT channel {channel}",
            )]
        else:
            return [TextContent(
                type="text",
                text="Failed to write to RTT.",
            )]

    elif name == "rtt_list_devices":
        devices = await asyncio.to_thread(rtt.list_devices)
        if devices:
            return [TextContent(
                type="text",
                text="Available J-Link devices:\n" + "\n".join(f"  - {d}" for d in devices),
            )]
        else:
            return [TextContent(
                type="text",
                text="No J-Link devices found. Make sure J-Link is connected.",
            )]

    elif name == "jlink_status":
        if not rtt.is_connected():
            return [TextContent(type="text", text="J-Link is not connected.")]

        # status() only reads cached fields; no blocking pylink call.
        status = rtt.status()
        return [TextContent(
            type="text",
            text=f"J-Link Status:\n"
                 f"  Connected: {status.connected}\n"
                 f"  RTT Started: {status.rtt_started}\n"
                 f"  Device: {status.device_name}\n"
                 f"  Serial: {status.serial}\n"
                 f"  Speed: {status.speed} kHz\n"
                 f"  Channel: {status.channel}\n"
                 f"  Buffer Size: {status.ring_buffer_size} entries",
        )]

    elif name == "rtt_clear":
        if not rtt.is_connected():
            return [TextContent(
                type="text",
                text="J-Link is not connected. Call jlink_connect first.",
            )]

        await asyncio.to_thread(rtt.clear_buffer)
        return [TextContent(type="text", text="RTT buffer cleared.")]

    else:
        return [TextContent(
            type="text",
            text=f"Unknown tool: {name}",
        )]

File: go/internal/tools/tools.go
// Package tools registers the nine RTT MCP tools on a server and dispatches
// them to the rttcore singleton. Tool names, argument schemas and result text
// are kept byte-for-byte compatible with the Python server so the VSCode
// extension and Claude Code see an identical surface.
package tools

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"

	"github.com/modelcontextprotocol/go-sdk/mcp"

	"rtt-mcp-server/internal/rttcore"
)

// Register installs every tool on the given server.
func Register(s *mcp.Server) {
	core := rttcore.Get()
	_ = core // referenced via rttcore.Get() in handlers to reflect singleton state

	mcp.AddTool(s,
		&mcp.Tool{
			Name:        "jlink_connect",
			Description: "Connect to J-Link debugger and start RTT monitoring. Must be called before reading or writing RTT data.",
		}, handleConnect)

	mcp.AddTool(s,
		&mcp.Tool{
			Name:        "jlink_disconnect",
			Description: "Disconnect from J-Link debugger and stop RTT monitoring.",
		}, handleDisconnect)

	mcp.AddTool(s,
		&mcp.Tool{
			Name:        "rtt_read",
			Description: "Read accumulated RTT data from the ring buffer. Data is continuously collected by the background monitor thread. Use this to get output from the target device.",
		}, handleRead)

	mcp.AddTool(s,
		&mcp.Tool{
			Name:        "rtt_read_log",
			Description: "Read the tail of the complete RTT log file. This is an independent broadcast sink (not drained by other clients), so it returns the full RTT output even while another client (e.g. the VSCode extension) is streaming via rtt_read. Prefer this over rtt_read when another consumer is active.",
		}, handleReadLog)

	mcp.AddTool(s,
		&mcp.Tool{
			Name:        "rtt_read_raw",
			Description: `Read new bytes from the broadcast log starting at a byte offset. Non-draining and multi-consumer safe: ideal for a continuous monitor that must coexist with other readers without stealing their data. Pass the returned next_offset as 'offset' on the next call; if the log rotated (next_offset > file size), pass offset=0. Returns JSON: {"data": "...", "next_offset": N}.`,
		}, handleReadRaw)

	mcp.AddTool(s,
		&mcp.Tool{
			Name:        "rtt_write",
			Description: "Write data to RTT down-buffer (host -> device). C-style escapes are interpreted so control bytes can be sent: \\r \\n \\t \\0 \\\\ and \\xNN (two hex digits) — e.g. pass AT\\r\\n to send 'AT' followed by CR/LF (4 bytes), not the 6 literal characters. A literal backslash is sent with \\\\. Unknown escapes are rejected. The target device must be running RTT with a down-buffer listener.",
		}, handleWrite)

	mcp.AddTool(s,
		&mcp.Tool{
			Name:        "rtt_list_devices",
			Description: "List available J-Link debuggers connected to the host.",
		}, handleListDevices)

	mcp.AddTool(s,
		&mcp.Tool{
			Name:        "rtt_list_supported_devices",
			Description: "List device names in the loaded J-Link device database (probe-less; works before connecting). Pass `query` for a case-insensitive substring filter. Use an exact returned name (e.g. STM32F407VE) as the `device` argument to jlink_connect.",
		}, handleListSupportedDevices)

	mcp.AddTool(s,
		&mcp.Tool{
			Name:        "rtt_check_device",
			Description: "Check whether a device name exists in the loaded J-Link device database (probe-less). Returns supported (yes/no) and the J-Link index.",
		}, handleCheckDevice)

	mcp.AddTool(s,
		&mcp.Tool{
			Name:        "jlink_status",
			Description: "Get current J-Link connection status and RTT buffer information.",
		}, handleStatus)

	mcp.AddTool(s,
		&mcp.Tool{
			Name:        "rtt_clear",
			Description: "Clear the RTT ring buffer. This does not clear the actual RTT buffers on the device.",
		}, handleClear)
}

// ---- input structs (pointer fields => optional in the derived schema) ----

type connectIn struct {
	Serial *string `json:"serial,omitempty"`
	Device *string `json:"device,omitempty"`
	Speed  *int    `json:"speed,omitempty"`
}

type readIn struct {
	Channel  *int `json:"channel,omitempty"`
	MaxBytes *int `json:"max_bytes,omitempty"`
}

type readLogIn struct {
	MaxBytes *int `json:"max_bytes,omitempty"`
}

type readRawIn struct {
	Offset   *int64 `json:"offset,omitempty"`
	MaxBytes *int   `json:"max_bytes,omitempty"`
}

type writeIn struct {
	Channel *int    `json:"channel,omitempty"`
	Data    *string `json:"data,omitempty"`
}

// ---- handlers ----

func handleConnect(ctx context.Context, req *mcp.CallToolRequest, in connectIn) (*mcp.CallToolResult, any, error) {
	c := rttcore.Get()
	defer c.TouchIdle() // refresh idle watchdog on any connect attempt that keeps the probe held
	serial := derefStr(in.Serial)
	device := derefStr(in.Device)
	speed := derefInt(in.Speed)

	// If already connected and the caller passed any parameter that conflicts
	// with the live connection, force a reconnect so the new value actually
	// takes effect. Empty / zero means "leave alone" (use default).
	if c.IsConnected() {
		st := c.Status()
		conflict := (device != "" && device != st.DeviceName) ||
			(speed != 0 && speed != st.Speed) ||
			(serial != "" && serial != st.Serial)
		if !conflict {
			return text(fmt.Sprintf(
				"Already connected to J-Link device '%s' (serial: %s, speed: %d kHz)\nRTT monitoring active on channel %d (shared connection)",
				st.DeviceName, st.Serial, st.Speed, st.Channel)), nil, nil
		}
		// Reconnect to honor the new parameters.
		c.Disconnect()
	}

	if err := c.Connect(serial, device, speed); err != nil {
		return text("Failed to connect to J-Link.\n\nError:\n" + err.Error()), nil, nil
	}
	st := c.Status()
	return text(fmt.Sprintf(
		"Connected to J-Link device '%s' (serial: %s, speed: %d kHz)\nRTT monitoring started on channel %d",
		st.DeviceName, st.Serial, st.Speed, st.Channel)), nil, nil
}

func handleDisconnect(ctx context.Context, req *mcp.CallToolRequest, in struct{}) (*mcp.CallToolResult, any, error) {
	c := rttcore.Get()
	// No TouchIdle: this releases the probe; the watchdog is stopped by Disconnect itself.
	if !c.IsConnected() {
		return text("J-Link is not connected."), nil, nil
	}
	c.Disconnect()
	return text("J-Link disconnected successfully."), nil, nil
}

func handleRead(ctx context.Context, req *mcp.CallToolRequest, in readIn) (*mcp.CallToolResult, any, error) {
	c := rttcore.Get()
	defer c.TouchIdle()
	// Transparently re-establish the probe if the idle watchdog released it, so a
	// read after an idle gap resumes the live stream instead of reporting "not
	// connected".
	if err := c.EnsureConnected(); err != nil {
		return text(err.Error()), nil, nil
	}
	data := c.Read(derefInt(in.MaxBytes))
	if data == "" {
		return text("(no RTT data)"), nil, nil
	}
	return text(data), nil, nil
}

func handleReadLog(ctx context.Context, req *mcp.CallToolRequest, in readLogIn) (*mcp.CallToolResult, any, error) {
	c := rttcore.Get()
	defer c.TouchIdle()
	data := c.ReadLogTail(derefInt(in.MaxBytes))
	if data == "" {
		return text("(no RTT log yet — connect first)"), nil, nil
	}
	return text(data), nil, nil
}

func handleReadRaw(ctx context.Context, req *mcp.CallToolRequest, in readRawIn) (*mcp.CallToolResult, any, error) {
	c := rttcore.Get()
	defer c.TouchIdle()
	var offset int64
	if in.Offset != nil {
		offset = *in.Offset
	}
	data, next := c.ReadLogRaw(offset, derefInt(in.MaxBytes))
	out, _ := json.Marshal(map[string]any{"data": data, "next_offset": next})
	return text(string(out)), nil, nil
}

func handleWrite(ctx context.Context, req *mcp.CallToolRequest, in writeIn) (*mcp.CallToolResult, any, error) {
	c := rttcore.Get()
	defer c.TouchIdle()
	// Transparently re-establish the probe if the idle watchdog released it: a
	// write that follows a successful connect must not fail with "not connected"
	// just because the idle window elapsed. This is the fix for the connect→write
	// state mismatch.
	if err := c.EnsureConnected(); err != nil {
		return text(err.Error()), nil, nil
	}
	if in.Data == nil || *in.Data == "" {
		return text("No data provided to write."), nil, nil
	}
	raw, err := unescape(*in.Data)
	if err != nil {
		return text("Invalid escape in data: " + err.Error() +
			"\nSupported: \\r \\n \\t \\0 \\\\ \\xNN"), nil, nil
	}
	channel := c.Config().Channel
	if in.Channel != nil {
		channel = *in.Channel
	}
	n := c.Write(channel, string(raw))
	if n < 0 {
		return text("Failed to write to RTT."), nil, nil
	}
	return text(fmt.Sprintf("Wrote %d bytes to RTT channel %d", n, channel)), nil, nil
}

func handleListDevices(ctx context.Context, req *mcp.CallToolRequest, in struct{}) (*mcp.CallToolResult, any, error) {
	c := rttcore.Get()
	devices := c.ListDevices()
	if len(devices) == 0 {
		return text("No J-Link devices found. Make sure J-Link is connected."), nil, nil
	}
	var b []byte
	b = append(b, "Available J-Link devices:\n"...)
	for _, d := range devices {
		b = append(b, "  - "...)
		b = append(b, d...)
		b = append(b, '\n')
	}
	return text(string(b)), nil, nil
}

type listSupportedIn struct {
	Query *string `json:"query,omitempty"`
	Limit *int    `json:"limit,omitempty"`
}

func handleListSupportedDevices(ctx context.Context, req *mcp.CallToolRequest, in listSupportedIn) (*mcp.CallToolResult, any, error) {
	c := rttcore.Get()
	n := c.SupportedDeviceCount()
	query := strings.ToLower(strings.TrimSpace(derefStr(in.Query)))
	limit := derefInt(in.Limit)
	if limit <= 0 {
		limit = 20000
	}
	var b []byte
	b = append(b, fmt.Sprintf("%d device(s) in the J-Link database.\n", n)...)
	matched := 0
	for i := 0; i < n && matched < limit; i++ {
		name := c.SupportedDeviceName(i)
		if name == "" {
			continue
		}
		if query != "" && !strings.Contains(strings.ToLower(name), query) {
			continue
		}
		b = append(b, name...)
		b = append(b, '\n')
		matched++
	}
	if matched >= limit && n > limit {
		b = append(b, fmt.Sprintf("... truncated at %d matches; pass a more specific `query`.\n", limit)...)
	}
	return text(string(b)), nil, nil
}

type checkDeviceIn struct {
	Device *string `json:"device,omitempty"`
}

func handleCheckDevice(ctx context.Context, req *mcp.CallToolRequest, in checkDeviceIn) (*mcp.CallToolResult, any, error) {
	c := rttcore.Get()
	dev := strings.TrimSpace(derefStr(in.Device))
	if dev == "" {
		return text("No device name given."), nil, nil
	}
	idx := c.SupportedDeviceIndex(dev)
	if idx > 0 {
		return text(fmt.Sprintf("Supported: %q is in the J-Link device database (index %d).", dev, idx)), nil, nil
	}
	return text(fmt.Sprintf("Not supported: %q is NOT in the J-Link device database. Use rtt_list_supported_devices to find the exact name J-Link expects.", dev)), nil, nil
}

func handleStatus(ctx context.Context, req *mcp.CallToolRequest, in struct{}) (*mcp.CallToolResult, any, error) {
	c := rttcore.Get()
	defer c.TouchIdle()
	// Transparently re-establish the probe if the idle watchdog released it, so a
	// status check after an idle gap reports a live connection instead of leaking
	// the internal "idle-disconnected" state. Consistent with read/write/clear.
	if err := c.EnsureConnected(); err != nil {
		return text(err.Error()), nil, nil
	}
	st := c.Status()
	return text(fmt.Sprintf(`J-Link Status:
  Connected: %v
  RTT Started: %v
  Device: %s
  Serial: %s
  Speed: %d kHz
  Channel: %d
  Buffer Size: %d entries
`, st.Connected, st.RTTStarted, st.DeviceName, st.Serial, st.Speed, st.Channel, st.RingBufferSize)), nil, nil
}

func handleClear(ctx context.Context, req *mcp.CallToolRequest, in struct{}) (*mcp.CallToolResult, any, error) {
	c := rttcore.Get()
	defer c.TouchIdle()
	// Transparently re-establish the probe if the idle watchdog released it, so a
	// clear after an idle gap works instead of reporting "not connected". This
	// keeps read/write/clear/status uniform: all four auto-reconnect on idle.
	if err := c.EnsureConnected(); err != nil {
		return text(err.Error()), nil, nil
	}
	c.Clear()
	return text("RTT buffer cleared."), nil, nil
}

// ---- helpers ----

func text(s string) *mcp.CallToolResult {
	return &mcp.CallToolResult{
		Content: []mcp.Content{&mcp.TextContent{Text: s}},
	}
}

func derefStr(p *string) string {
	if p == nil {
		return ""
	}
	return *p
}

func derefInt(p *int) int {
	if p == nil {
		return 0
	}
	return *p
}

// unescape interprets C-style escape sequences in s and returns the raw bytes.
// Supported: \r \n \t \0 \\ \xNN (exactly two hex digits). Unknown or malformed
// escapes return an error so a typo can't silently send the wrong bytes — a
// literal backslash must be written as \\. This lets a caller send AT\r\n and
// get 'A','T',CR,LF (4 bytes) instead of the 6 literal characters.
func unescape(s string) ([]byte, error) {
	b := []byte(s)
	var out []byte
	for i := 0; i < len(b); {
		if b[i] != '\\' {
			out = append(out, b[i])
			i++
			continue
		}
		if i+1 >= len(b) {
			return nil, fmt.Errorf("dangling '\\' at end of input")
		}
		switch b[i+1] {
		case 'r':
			out = append(out, '\r')
			i += 2
		case 'n':
			out = append(out, '\n')
			i += 2
		case 't':
			out = append(out, '\t')
			i += 2
		case '0':
			out = append(out, 0)
			i += 2
		case '\\':
			out = append(out, '\\')
			i += 2
		case 'x':
			if i+3 >= len(b) {
				return nil, fmt.Errorf("\\x must be followed by exactly two hex digits")
			}
			hi, ok1 := hexVal(b[i+2])
			lo, ok2 := hexVal(b[i+3])
			if !ok1 || !ok2 {
				return nil, fmt.Errorf("invalid hex digit in %q", s[i:i+4])
			}
			out = append(out, byte(hi<<4|lo))
			i += 4
		default:
			return nil, fmt.Errorf("unknown escape sequence \\%c", b[i+1])
		}
	}
	return out, nil
}

// hexVal maps an ASCII hex digit to its 0–15 value (ok=false if not hex).
func hexVal(b byte) (int, bool) {
	switch {
	case b >= '0' && b <= '9':
		return int(b - '0'), true
	case b >= 'a' && b <= 'f':
		return int(b-'a') + 10, true
	case b >= 'A' && b <= 'F':
		return int(b-'A') + 10, true
	}
	return 0, false
}
